fix: token decimals fetched from the chain

tommasini opened the pull request:
## **Description**
The problem was that the user was able to insert wrong the token decimals values, and with that would get the wrong tokens amounts in their wallet.

* Added a new state to lock the text inputs of symbol and decimals when the symbol
* Decimals and names are fetchable from the chain
* Reset all the states when the address text input changed when an address does not have 42 chars
* Added styling to address the value of symbols and decimals text input when fetched from the chain

It was also added to the wallet_watchAsset JSON rpc method
* overrides token decimals and symbol if they aare fetched from the chain


## *Note for review*
*  On RPCMethodMiddleware  file `wallet_watchAsset` method
 * On the 207 line validateCustomTokenAddress of AddCustomToken  

## **Manual testing steps**
Scenario 1:
_1. Step1: Go to import token on wallet view
_2. Step2: change tab to custom token
_3. Step3: add `0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599` (WBTC) as token address
_4. Step4: It should block all the symbol and decimals field 


Scenario 2
 Step1: Go to in app browser
 Step2: Go to coingecko website
 Step3: Browse WBTC
 Step4: look for the MM icon 
 Step5: Add it via coin gecko
 Step6: It should work as expected

## **Screenshots/Recordings**
* importing token on custom token screen
http://recordit.co/ENT3D6Y0wn

* import token on coin gecko
https://recordit.co/Ikwc5adYZn

* import token on test dapp with wrong symbol and decimals (WBTC contract)
https://recordit.co/zoixJsK4Sc

* import WBTC on polygon network
https://recordit.co/zTtxIthF4C

## **E2E build**
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/1d5eb4d0-faff-456b-b66b-6e55718f4cdf


## **Related issues**

_Fixes #

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [ ] I’ve indicated what issue this PR is linked to: Fixes #???
- [ ] I’ve included tests if applicable.
- [ ] I’ve documented any added code.
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

tommasini committed:
added a new state to lock the text inputs of symbol and decimals when the symbol, decimals and name are fetchable from the chain, also reseted all the state when the address text input change when an address does not have 42 chars, added styling to address the value of symbols and decimals text input when fetched from the chain

tommasini committed:
Merge branch 'main' into fix/1307-check-decimals-on-chain

tommasini committed:
overrides symbol and decimals of a token if the address fetch that values from the chain on wallet_watchAsset jsonRPC method

Cal-L commented on app/components/UI/AddCustomToken/index.js:
```diff
@@ -155,8 +160,39 @@ export default class AddCustomToken extends PureComponent {
+        this.setState({ isSymbolAndDecimalEditable: true });
+      }
+    } else {
+      this.setState({
```
Add comment that we're clearing other fields when changing address

Cal-L commented on app/components/UI/AddCustomToken/index.js:
```diff
@@ -155,8 +160,39 @@ export default class AddCustomToken extends PureComponent {
-  onAddressChange = (address) => {
-    this.setState({ address });
+  onAddressChange = async (address) => {
+    this.setState({ address, isSymbolAndDecimalEditable: true });
```
Move this logic to the else of if (validated)

Cal-L commented on app/components/UI/AddCustomToken/index.js:
```diff
@@ -167,36 +203,30 @@ export default class AddCustomToken extends PureComponent {
-
-  validateCustomTokenAddress = async () => {
+  validateCustomTokenAddress = async (address) => {
+    const tokenAddress = address || this.state.address;
```
Pass this.state.address in instead

legobeat reviewed (dismissed):
Existing test failing: https://github.com/MetaMask/metamask-mobile/actions/runs/6566850147/job/17838459943?pr=7540

tommasini committed:
address review, added isValidAddress to rpc method

tommasini committed:
update snapshot

tommasini commented:
@legobeat Addressed the test failure, the unit test would be a blocker for you? I'm having a hard time with this unit test and I feel the logic that will be tested is the override logic due to mocking the other results. Let me know your thoughts <3

@Cal-L Addressed all your reviews thanks <3

tommasini committed:
added tests and extracted the rpc method to a new single file for better reading

Cal-L committed:
Merge branch 'main' into fix/1307-check-decimals-on-chain

Andepande reviewed (approved):
QA Passed
